### PR TITLE
Add SFT vision COCO captions dataset and config support

### PR DESCRIPTION
**Pull request by subawocit**
# Description

This PR adds support for COCO-caption dataset for Supervised Fine-Tuning.

### Files and Changes
- `src/maxtext/input_pipeline/hf_data_processing.py`: Added `add_default_prompt_if_missing` to automatically inject `default_prompt` when the specified prompt column is missing in dataset features (e.g. COCO captioning).
- `src/maxtext/configs/base.yml` and `src/maxtext/configs/types.py`: Registered `default_prompt`.
- `src/maxtext/configs/post_train/sft-vision-coco-captions.yml`: Added configuration for COCO-caption dataset.
- `tests/unit/hf_data_processing_test.py`: Added `test_add_default_prompt_if_missing` unit test verifying prompt injection when missing and preservation when present.

# Tests

Tested end-to-end multimodal SFT and decode pipeline on Gemma 3 4B:

### 1. Pre-SFT Baseline Inference
```bash
python3 -m maxtext.inference.decode \
    src/maxtext/configs/base.yml \
    model_name=gemma3-4b \
    tokenizer_path=google/gemma-3-4b-it \
    tokenizer_type=huggingface \
    load_parameters_path=gs://yuchenhou-maxtext-logs/checkpoints/gemma3-4b-processor/0/items \
    per_device_batch_size=1 \
    run_name=gemma3_4b_coco_pre_sft \
    max_prefill_predict_length=1024 \
    max_target_length=2048 \
    steps=1 \
    async_checkpointing=false \
    scan_layers=false \
    use_multimodal=true \
    prompt='Describe the image concisely' \
    image_path=tests/assets/test_image.jpg \
    attention=dot_product \
    skip_jax_distributed_system=true
```
**Output:**
```
Input `<start_of_turn>user
<start_of_image>Describe the image concisely<end_of_turn>
<start_of_turn>model
` -> `Here's a concise description of the image:

The image shows a panoramic view of Seattle, Washington. The iconic Space Needle dominates the skyline, with numerous other buildings visible. In the distance, snow-capped mountains are partially obscured by clouds.`
```

---

### 2. Vision SFT Training on COCO
```bash
python3 -m maxtext.trainers.post_train.sft.train_sft_native \
    src/maxtext/configs/post_train/sft-vision-coco-captions.yml \
    model_name=gemma3-4b \
    tokenizer_path=google/gemma-3-4b-it \
    tokenizer_type=huggingface \
    load_parameters_path=gs://yuchenhou-maxtext-logs/checkpoints/gemma3-4b-processor/0/items \
    base_output_directory=gs://yuchenhou-maxtext-logs/gemma3-4b/multimodal/sft \
    run_name=gemma3_4b_sft_coco \
    per_device_batch_size=1 \
    max_prefill_predict_length=1024 \
    max_target_length=2048 \
    steps=101 \
    checkpoint_period=20 \
    scan_layers=false \
    async_checkpointing=false \
    save_checkpoint_on_completion=true \
    log_period=1 \
    eval_interval=10 \
    eval_steps=2 \
    gcs_metrics=true \
    float32_qk_product=true \
    float32_logits=true \
    attention=dot_product \
    sharding_tolerance=0.05 \
    checkpoint_storage_use_zarr3=false \
    checkpoint_storage_use_ocdbt=false \
    enable_single_controller=false \
    grain_worker_count=0
```

Verified locally that both train and val losses are decreasing and no errors detected.

---

### 3. Post-SFT Inference Verification
```bash
python3 -m maxtext.inference.decode \
    src/maxtext/configs/base.yml \
    model_name=gemma3-4b \
    tokenizer_path=google/gemma-3-4b-it \
    tokenizer_type=huggingface \
    load_parameters_path=gs://yuchenhou-maxtext-logs/gemma3-4b/multimodal/sft/gemma3_4b_sft_coco/checkpoints/100/items \
    per_device_batch_size=1 \
    run_name=gemma3_4b_coco_post_sft \
    max_prefill_predict_length=1024 \
    max_target_length=1600 \
    steps=1 \
    async_checkpointing=false \
    scan_layers=false \
    use_multimodal=true \
    prompt='Describe the image concisely.' \
    image_path=tests/assets/test_image.jpg \
    attention=dot_product \
    skip_jax_distributed_system=true
```
**Output:**
```
Input `<start_of_turn>user
<start_of_image>Describe the image concisely.<end_of_turn>
<start_of_turn>model
` -> `A city skyline with a snow covered mountain`
```

---

### 4. Unit Tests
```bash
JAX_PLATFORMS=cpu pytest tests/unit/hf_data_processing_test.py
```
**Output:**
```
=========================== 4 passed, 1 warning in 92.40s (0:01:32) ===========================
```

---

# Checklist
Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/input_pipeline/hf_data_processing.py [hunk 1]
```diff
@@ -42,6 +42,18 @@ def _get_pad_id(tokenizer):
...
 
 
+def add_default_prompt_if_missing(dataset, text_columns, config):
+  """Inject default prompt if prompt column is missing in dataset (e.g. COCO captioning)."""
+  prompt_col = text_columns[0]
+  # Check if features is None (e.g. streaming IterableDataset) or prompt column is absent
+  features = getattr(dataset, "features", None) or getattr(dataset, "column_names", None)
+  # If so, populate each example with default_prompt
+  if features is None or prompt_col not in features:
+    default_prompt = getattr(config, "default_prompt", "Describe this image")
```
Comment (igorts-git) [inline src/maxtext/input_pipeline/hf_data_processing.py]: Do we still need `getattr` here now that we added `default_prompt`?

File: tests/unit/hf_data_processing_test.py [hunk 2]
```diff
@@ -123,6 +123,34 @@ def get_first_batch(iterator):
     self.assertTrue((train_batch1["inputs"] == train_batch2["inputs"]).all())  # pytype: disable=unsupported-operands
     self.assertTrue((train_batch1["targets"] == train_batch2["targets"]).all())  # pytype: disable=unsupported-operands
 
+  def test_add_default_prompt_if_missing(self):
+    import datasets  # pylint: disable=import-outside-toplevel
```
Comment (igorts-git) [inline tests/unit/hf_data_processing_test.py]: Do we need this here? looks like the AI agent is lazy and does not want to put imports at the top of the file.

**Review by igorts-git — APPROVED** [review]

File: src/maxtext/input_pipeline/hf_data_processing.py (hunk 1 above)
Comment (subawocit) [reply src/maxtext/input_pipeline/hf_data_processing.py]: Good catch - I removed this `getattr`

File: tests/unit/hf_data_processing_test.py (hunk 2 above)
Comment (subawocit) [reply tests/unit/hf_data_processing_test.py]: Agree - moved this import up

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -1403,6 +1403,10 @@ class DatasetGeneral(BaseModel):
   train_image_column: str | list[str] = Field("image", description="Column name(s) for images in the training data.")
   eval_data_columns: list[str] = Field(["text"], description="Column(s) to use from the evaluation data.")
   eval_image_column: str | list[str] = Field("image", description="Column name(s) for images in evaluation data.")
+  default_prompt: str = Field(
+      "Describe this image",
```
Comment (aireenmei) [inline src/maxtext/configs/types.py]: I suggest we set the default value to empty, set it only in the yaml for coco dataset. This is to prevent some user error in setting column name, and silently result in using default prompt

**Review by aireenmei — APPROVED** [review]
LGTM in general, with a minor suggestion

File: src/maxtext/configs/types.py (hunk 3 above)
Comment (subawocit) [reply src/maxtext/configs/types.py]: Good idea - edited
